openml/datasets/functions.py

```diff
@@ -418,22 +416,62 @@ def create_dataset(name, description, creator, contributor, collection_date,
         'data': data
     }
 
+    # Determine arff format from the dataset
+    if isinstance(data, list):
+        if isinstance(data[0], list):
+            d_format = 'arff'
+        elif isinstance(data[0], dict):
+            d_format = 'sparse_arff'
+        else:
+            raise ValueError('Illegal dataset value, '
+                             'only list of lists/dicts is supported')
+    elif isinstance(data, np.ndarray):
+        d_format = 'arff'
+    elif isinstance(data, coo_matrix):
+        d_format = 'sparse_arff'
+    else:
+        raise ValueError('Illegal dataset value, '
+                         'please check the function documentation')
+
     # serializes the arff dataset object and returns a string
     arff_dataset = arff.dumps(arff_object)
     try:
         # check if arff is valid
         decoder = arff.ArffDecoder()
-        decoder.decode(arff_dataset, encode_nominal=True)
+        if d_format == 'arff':
```

Comment:
It might be shorter to write:

```python
return_type = arff.COO if d_format == 'sparse_arff' else arff.DENSE
decoder.decode(arff_dataset, encode_nominal=True, return_type=return_type)
```